### PR TITLE
JDK-8282696: Add constructors taking a cause to InvalidObjectException and InvalidClassException

### PR DESCRIPTION
**Pull request by jddarcy**
Occasionally in core-libs we've discussed whether or not to do a pass over the exception classes and proactively add any of four missing convention constructors per java.lang.Throwable (no-arg, string, cause, cause and string). Last time this came up, we decided a wide-scale effort wasn't worthwhile.

Prompted by some other recent work, I decided to take a quick look at the dual-approach: grep for calls to initCause in java.base and seeing which exception classes repeated were initCaused. Those exception classes are good candidates for cause-taking constructors.

Two such exception classes area InvalidObjectException and InvalidClassException, along with their superclass ObjectStreamException.

Please also review the CSR: https://bugs.openjdk.java.net/browse/JDK-8282697

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8282696](https://bugs.openjdk.java.net/browse/JDK-8282696): Add constructors taking a cause to InvalidObjectException and InvalidClassException
 * [JDK-8282697](https://bugs.openjdk.java.net/browse/JDK-8282697): Add constructors take a cause to InvalidObjectException and InvalidClassException (**CSR**)

### Reviewers
 * [Lance Andersen](https://openjdk.java.net/census#lancea) (@LanceAndersen - **Reviewer**) ⚠️ Review applies to 80f97dcc9803c590cf4be20b6750d274d2d68d36

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7711/head:pull/7711` \
`$ git checkout pull/7711`

Update a local copy of the PR: \
`$ git checkout pull/7711` \
`$ git pull https://git.openjdk.java.net/jdk pull/7711/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7711`

View PR using the GUI difftool: \
`$ git pr show -t 7711`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7711.diff">https://git.openjdk.java.net/jdk/pull/7711.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/io/InvalidObjectException.java [hunk 1]
```diff
@@ -48,4 +48,18 @@
...
+     * reason and cause.
+     *
+     * @param reason Detailed message explaining the reason for the failure.
+     * @param cause the cause
+     *
+     * @see ObjectInputValidation
+     * @since 19
+     */
+    public  InvalidObjectException(String reason, Throwable cause) {
+        super(reason);
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/io/InvalidObjectException.java:63]: Should this be: `super(reason, cause)`?

File: src/java.base/share/classes/java/io/ObjectInputStream.java [hunk 2]
```diff
@@ -2221,9 +2213,9 @@ private Object readArray(boolean unshared) throws IOException {
                 Enum<?> en = Enum.valueOf((Class)cl, name);
                 result = en;
             } catch (IllegalArgumentException ex) {
-                throw (IOException) new InvalidObjectException(
-                    "enum constant " + name + " does not exist in " +
-                    cl).initCause(ex);
+                throw new InvalidObjectException("enum constant " +
+                                                 name + " does not exist in " +
+                                                 cl, ex);
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/io/ObjectInputStream.java:2218]: Can this be reflowed to be on two lines instead of 3?  Your call.

**Review by RogerRiggs — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/java/io/InvalidObjectException.java (hunk 1 above)
Comment (jddarcy) [reply src/java.base/share/classes/java/io/InvalidObjectException.java:63]: Yes; good catch -- will fix in the next push. Thanks.

**Review by LanceAndersen — APPROVED** [review]

File: src/java.base/share/classes/java/io/InvalidObjectException.java [hunk 3]
```diff
@@ -48,4 +48,18 @@
...
+     * Constructs an {@code InvalidObjectException} with the given
+     * reason and cause.
+     *
+     * @param reason Detailed message explaining the reason for the failure.
+     * @param cause the cause
+     *
+     * @see ObjectInputValidation
+     * @since 19
+     */
+    public  InvalidObjectException(String reason, Throwable cause) {
```
Comment (turbanoff) [inline src/java.base/share/classes/java/io/InvalidObjectException.java:62]: strange double space after `public`

File: src/java.base/share/classes/java/io/InvalidObjectException.java [hunk 4]
```diff
@@ -48,4 +48,18 @@
     public  InvalidObjectException(String reason) {
```
Comment (turbanoff) [inline src/java.base/share/classes/java/io/InvalidObjectException.java:48]:
```suggestion
    public InvalidObjectException(String reason) {
```

File: src/java.base/share/classes/java/io/InvalidObjectException.java (hunk 3 above)
Comment (jddarcy) [reply src/java.base/share/classes/java/io/InvalidObjectException.java:62]: Fixed in subsequent push; thanks.

Comment (turbanoff) [reply src/java.base/share/classes/java/io/InvalidObjectException.java:62]: Hm. I see it still there)

**Review by LanceAndersen — APPROVED** [review]

**Pull request closed without merging**